### PR TITLE
feat: List v2 - Enable ListWidgetV2 APIs. 

### LINKED ISSUE #9568: [Bug]: cannot reach widgets inside a list if the list has a "OnListItemClick" property specified.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When the list has some widgets inside, and something specified in the "OnListItemClick" property. Now you cannot select the widgets inside the list. When you attempt to, the OnListItemClick gets called instead. :(

Uploading Screen Recording 2021-12-04 at 9.27.37 PM.mov…

### Steps To Reproduce

1. I have created an app to reproduce this: https://app.appsmith.com/applications/61ab8e37d3e4214454c23c86/pages/61ab8e37d3e4214454c23c88/edit
2. see there are 2 lists: List1 and List2
3. List1 has OnListItemClick property to open some modal. List2 has no OnListItemClick event specified. 
4. When you try to access some widget inside List1, the modal opens and the property pane also shows the properties of the modal. You cannot anymore access the properties of the widgets inside the list unless you remove the OnListItemClick property and reset it to "no action"
5. You can do all this with List2, I'm guessing because it has no OnListItemClick specified. 

### Environment

Production

### Version

Cloud

### LINKED ISSUE #17480: Task: List v2 - Enable ListWidgetV2 APIs.
Add the following APIs to the List widget to help users access data from the List widget.

```
listData
  Singleton - listDataItem

item -> from listData
row  -> from the list row i.e widgets

selectedItemIndex       -> selectedRowIndex
selectedItem            -> selectedItem 
selectedViewItemIndex   -> selectedRowViewIndex (Internal)
currentSelectedViewItem -> selectedRow
currentViewItems        -> currentViewRows
                        -> triggeredRowIndex
                        -> triggeredRowViewIndex (Internal)
                        -> triggeredRow
onListItemClick         -> onRowClick
currentItem             -> currentItem
currentIndex            -> currentIndex
currentRow              -> currentRow

level (TBD)

```

**NB:** The following should work for Pagination and Infinite Scrolling

Front logo [Front conversations](https://app.frontapp.com/open/top_5zhjg)

### PR DESCRIPTION
**Pull request by Tooluloope**
## Description

-  Added `onRowClick` and `onRowClickCapture` to the container just before the widget is created to help trigger the `onRowClick` and `triggerRowClick` respectively. `onRowClick` updates `triggeredRowIndex`, `triggeredRow`, `selectedRowIndex` and `selectedRow`.   `onRowClickCapture` only updates `triggeredRow` and `triggeredRowIndex`.

I completely removed `selectedViewItemIndex` and `triggeredViewItemIndex`, since I didn't need them.

Migrated the following

onListItemClick         -> onRowClick
selectedItemIndex       -> selectedRowIndex
currentViewItems        -> currentViewRows

For `selectedRow` and `triggeredRow` caching was required, so I kept the metaWidgets of these rows in the Datatree but filtered it from view using `this.currViewMetaWidgetIds` so it doesn't duplicate if the selected or triggeredRow is in view.

Fixes #17480 

## Type of change

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- Cypress

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/client/src/widgets/ListWidgetV2/widget/index.tsx [hunk 1]
```diff
@@ -639,7 +736,24 @@ class ListWidget extends BaseWidget<ListWidgetProps, WidgetState> {
...
+          const calculatedSelectedRowIndex = this.getContainerIndex(index);
+          return {
+            ...container,
+            selected: selectedRowIndex === calculatedSelectedRowIndex,
+            onClick: (e: React.MouseEvent<HTMLElement>) => {
+              e.stopPropagation();
+              this.onRowClick(
+                calculatedSelectedRowIndex,
+                index,
+                this.props.onRowClick,
```
Comment (ashit-rath) [inline app/client/src/widgets/ListWidgetV2/widget/index.tsx:749]: I don't think we need to pass this is an argument. the `this.onRowClick` can directly call `this.props.onRowClick`

File: app/client/src/widgets/ListWidgetV2/widget/index.tsx [hunk 2]
```diff
@@ -571,33 +582,104 @@ class ListWidget extends BaseWidget<ListWidgetProps, WidgetState> {
...
+
+  resetSelectedRowIndexMeta = () => {
+    this.props.updateWidgetMetaProperty("selectedRowIndex", -1);
+  };
+  resetTriggeredRowIndexMeta = () => {
+    this.props.updateWidgetMetaProperty("triggeredRowIndex", -1);
   };
 
+  getGridGap = () =>
+    this.props.gridGap && this.props.gridGap >= -8 ? this.props.gridGap : 0;
```
Comment (ashit-rath) [inline app/client/src/widgets/ListWidgetV2/widget/index.tsx:681]: Let's move `-8` to a const

File: app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts [hunk 3]
```diff
@@ -215,6 +225,9 @@ class MetaWidgetGenerator {
     this.serverSidePagination = options.serverSidePagination;
     this.templateBottomRow = options.templateBottomRow;
     this.widgetName = options.widgetName;
+    this.selectedRowIndex = options.selectedRowIndex;
```
Comment (ashit-rath) [inline app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts:228]: Sort these alphabetically

File: app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts [hunk 4]
```diff
@@ -182,6 +190,8 @@ class MetaWidgetGenerator {
     this.onVirtualListScroll = props.onVirtualListScroll;
     this.pageNo = 1;
     this.pageSize = 0;
+    this.selectedRowIndex = -1;
+    this.triggeredRowIndex = -1;
```
Comment (ashit-rath) [inline app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts:194]: Sort alphabetically

File: app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts [hunk 5]
```diff
@@ -153,6 +159,8 @@ class MetaWidgetGenerator {
   private onVirtualListScroll: ConstructorProps["onVirtualListScroll"];
   private pageNo?: number;
   private pageSize?: number;
+  private selectedRowIndex: number;
```
Comment (ashit-rath) [inline app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts:162]: Sort alphabetically

File: app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts [hunk 6]
```diff
@@ -310,8 +323,10 @@ class MetaWidgetGenerator {
 
       indices.forEach((rowIndex) => {
         const index = startIndex + rowIndex;
```
Comment (ashit-rath) [inline app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts:325]: I think we have to do a refactor here.
rename 
index -> rowIndex
rowIndex -> viewIndex.

We will do it in a separate PR
What do you think?

File: app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts [hunk 7]
```diff
@@ -352,24 +367,116 @@ class MetaWidgetGenerator {
...
+      metaWidgets: {
+        ...metaWidgets,
+        ...notInViewSelectedRowMetaWidgets,
+        ...notInViewTriggeredRowMetaWidgets,
+      },
       removedMetaWidgetIds,
     };
   };
 
+  private filterGeneratedWidgetNotInView = ({
```
Comment (ashit-rath) [inline app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts:394]: I think the whole logic here can be simplified here.

Let's say we are on page 1 and we select a row (Row 2).
As that row is present in the current view, we are able to show the `selectedRow` 

Now when we move to page 2, we remove all the widgets from page 1 and in turn remove the row widgets of selectedRow.

What I think we can probably do is when we return the `removedMetaWidgetIds`, we just simply not include the widgetIds that are part of the selected row. So basically we are preserving them.

Right now with a quick look what I think we are doing is re-generating the row's widgets which can be avoided.

Let me know if this makes sense and if I overlooked some key think that resulted in the current direction you took.

**Comment by ashit-rath** [comment]
Screenshot 2022-11-18 at 5 47 09 PM

Also saw this `undefined` in autocomplete


Screenshot 2022-11-18 at 7 23 48 PM

There was also this `items` which has this `undefined` value. Not sure if this was from before but do check this out.

File: app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts [hunk 8]
```diff
@@ -245,13 +258,13 @@ class MetaWidgetGenerator {
     }
   };
 
-  didMount = () => {
```
[the remark this answers is not in the document]
Comment (Tooluloope) [reply app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts:248]: I had to go inside `MetaWidgetGenerator` from `widget/index` to understand what was unmounting. 
Since it's a public method, I felt the Variable name should be clear.

File: app/client/src/widgets/ListWidgetV2/widget/index.tsx [hunk 9]
```diff
@@ -190,6 +192,13 @@ class ListWidget extends BaseWidget<ListWidgetProps, WidgetState> {
       }
     }
 
+    if (!isEqual(this.props.primaryKeys, prevProps.primaryKeys)) {
+      this.resetSelectedRowIndexMeta();
```
[the remark this answers is not in the document]
Comment (Tooluloope) [reply app/client/src/widgets/ListWidgetV2/widget/index.tsx:196]: Just better naming.
I'll remove it.

File: app/client/src/widgets/ListWidgetV2/widget/index.tsx [hunk 10]
```diff
@@ -571,33 +582,104 @@ class ListWidget extends BaseWidget<ListWidgetProps, WidgetState> {
...
-    });
+  updateSelectedRowIndexMeta = (rowIndex: number) => {
+    const { selectedRowIndex } = this.props;
+
+    if (rowIndex === selectedRowIndex) {
+      this.resetSelectedRowIndexMeta();
+      return;
+    }
+
+    this.props.updateWidgetMetaProperty("selectedRowIndex", rowIndex);
```
[the remark this answers is not in the document]
Comment (Tooluloope) [reply app/client/src/widgets/ListWidgetV2/widget/index.tsx:615]: It'll be more difficult for someone else to read at a glance.

File: app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts (hunk 6 above)
Comment (Tooluloope) [reply app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts:325]: I'd make a PR after this to fix that.
I think it's better to do that immediately than later

File: app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (ashit-rath) [reply app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts:248]: All of these methods were introduced similar to lifecycle event. Right now it just deals with virtualizer but in future it would do a lot more (Like the unmount is resetting the cache along with the virtualizer)

In terms of what it is mounting and unmounting is completely internal to the `MetaWidgetGenerator` and shouldn't be a concern of the caller as `virtualizer` is a concept that the List widget is unaware of and having to call a method like `unmountVirtualizer` would pose much more questions regarding the existence and usability of the `virtualizer`.

If this was something that is being called multiple times (mount and unmount is called only once), then we should have thought of meaningful names.

File: app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts (hunk 6 above)
Comment (Tooluloope) [reply app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts:328]: https://github.com/appsmithorg/appsmith/issues/18287

**Comment by Tooluloope** [comment]
> Screenshot 2022-11-18 at 5 47 09 PM
> 
> Also saw this `undefined` in autocomplete
> 
> Screenshot 2022-11-18 at 7 23 48 PM
> 
> There was also this `items` which has this `undefined` value. Not sure if this was from before but do check this out.

You can remove this in your PR https://github.com/appsmithorg/appsmith/pull/18278.

This is similar to currentViewItems, except that currentViewItems is just for the current view, and yes, this was from V1.

It'd only be used if we show all the items on all pages in the List widget.

File: app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts [hunk 11]
```diff
@@ -938,6 +1030,15 @@ class MetaWidgetGenerator {
     this.setCache(updatedCache);
   };
 
+  private clearCacheByRow = (row: string) => {
```
Comment (ashit-rath) [inline app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts:1033]: Double check if this is needed after the discussed changes

File: app/client/src/widgets/ListWidgetV2/widget/index.tsx [hunk 12]
```diff
@@ -540,64 +604,126 @@ class ListWidget extends BaseWidget<ListWidgetProps, WidgetState> {
...
-  getGridGap = () =>
-    this.props.gridGap && this.props.gridGap >= -8 ? this.props.gridGap : 0;
-
-  updateActions = (children: DSLWidget[]) => {
-    return children.map((child: DSLWidget, index) => {
-      return {
-        ...child,
-        onClickCapture: () =>
-          this.onItemClick(index, this.props.onListItemClick),
-        selected: this.props.selectedItemIndex === index,
```
Comment (ashit-rath) [inline app/client/src/widgets/ListWidgetV2/widget/index.tsx:595]: We need to add `focused` property to the container widget. It makes sure that the z-index of the first container is higher.

File: app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts [hunk 13]
```diff
@@ -113,6 +115,10 @@ enum MODIFICATION_TYPE {
 
 const ROOT_CONTAINER_PARENT_KEY = "__$ROOT_CONTAINER_PARENT$__";
 const ROOT_ROW_KEY = "__$ROOT_KEY$__";
+const SELECTED_ROW_KEY = "__$SELECTED_ROW_KEY$__";
+const TRIGGER_ROW_KEY = "__$TRIGGER_ROW_KEY$__";
+const SpecialKeys = [SELECTED_ROW_KEY, TRIGGER_ROW_KEY];
```
[the remark this answers is not in the document]
Comment (Tooluloope) [reply app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts]: Should I still use the Cache for this?

[the remark this answers is not in the document]
Comment (Tooluloope) [reply app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts]: I ended up using a Class property to cache it.
And since we might be caching some other rows in the future, sending this might be tedious.

File: app/client/src/widgets/ListWidgetV2/widget/index.tsx (hunk 12 above)
Comment (Tooluloope) [reply app/client/src/widgets/ListWidgetV2/widget/index.tsx:595]: I didn't get this part.
Do you mean I should just add `focused` for the first container i.e
```
focused: rowIndex === 0
```

Also, would the focused property be for edit mode alone?

File: app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts [hunk 14]
```diff
@@ -366,21 +377,88 @@ class MetaWidgetGenerator {
     };
   };
 
+  private getMetaWidgetIdsInCachedRows = () => {
+    const templateWidget = this.currTemplateWidgets?.[this.containerWidgetId];
+    const cachedMetaWidgetIds: string[] = [];
+    const removedCachedMetaWidgetIds: string[] = [];
+
+    this.cachedRows.prevCachedRows.forEach((key) => {
```
[the remark this answers is not in the document]
Comment (ashit-rath) [reply app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts:385]: >How do we clear [1, 2]

Wouldn't 1 and 2 be present in the `prevViewMetaWidgetIds` and get automatically removed?

[the remark this answers is not in the document]
Comment (Tooluloope) [reply app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts:385]: Yeah, since it's not part of currViewMetaWidgetIds, it won't be in prevViewMetaWidgetIds.

This is because
1. currViewMetaWidgetIds maintains the widget that is in view + overscan. So adding it here would cache and let the user see it in `currentViewItems`, which is not what we want.

[the remark this answers is not in the document]
Comment (ashit-rath) [reply app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts:385]: Okay, makes sense!

[the remark this answers is not in the document]
Comment (Tooluloope) [reply app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts]: Got it

**Review by ashit-rath — APPROVED** [review]

**Comment by chandannkumar** [comment]
Tested this PR
- Tested bug #9568 and LGTM
- Tested all List v2 API's enabled, mentioned on #17480
- Tested onRowClick Events in List v2
